Add workspaces to MooseServer by registering document resources

brandonlangley opened the pull request:
refs #22766

https://github.com/user-attachments/assets/831ace5c-91a8-4bbc-b4b4-025826403639

brandonlangley committed:
Fix MooseServer to return early from diagnostics if no parse tree

refs #22766

brandonlangley committed:
Fix MooseServer to return early from plotting if app build failed

refs #22766

brandonlangley committed:
Simplify MooseServerTest by removing unneeded string constructors

refs #22766

brandonlangley committed:
Add MooseServer workspaces by registering document resource files

refs #22766

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/3214838/), step Docs: sync website on 77bfaa2 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/31522/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3215018/), step Versioner verify on 85c4524 wanted to post the following:

### Versioner templates

Found 14 templates, 0 failed


### Versioner influential files

Found 58 influential files, 4 changed, 0 added, 0 removed

| package   | status   | file                                    |
|-----------|----------|-----------------------------------------|
| wasp      | CHANGE   | conda/wasp/meta.yaml                    |
| wasp      | CHANGE   | framework/contrib/wasp                  |
| moose-dev | CHANGE   | conda/moose-dev/conda_build_config.yaml |
| moose-dev | CHANGE   | conda/moose-dev/meta.yaml               |

### Versioner versions

Found 9 packages, 2 changed, 0 failed

| package   | status   | hash    | to hash   | version            | to version         |
|-----------|----------|---------|-----------|--------------------|--------------------|
| wasp      | CHANGE   | 41363d4 | 7d0da85   | 2025.09.04 build 0 | 2025.09.14 build 0 |
| moose-dev | CHANGE   | 1dc2fb4 | 74ee690   | 2025.09.04         | 2025.09.14         |

brandonlangley commented:
@loganharbour -

Workspace updates (i.e., resource files of a document) is an [LSP feature](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_didChangeWatchedFiles) the MooseServer previously did not support.

I recently updated the base ServerImpl class in WASP to handle this and register resource files for the client to watch.

This PR updates the MooseServer to add resource files that will get registered for the client to watch for any changes.

These are `!include` files and `FileName`, `FileNameNoExtension`, `MeshFileName`, `MatrixFileName` parameter values.

The client notifies the server when any watched file is modified, and the server then reprocesses the base document.

I added a test of a scenario modifying `!include` and `FileName` external resource files to verify diagnostics changes.

This also [fixes](https://github.com/idaholab/moose/pull/31522/commits/aafac09ed20101d77fa1e169d946eb67e0d290db) the MooseServer empty document crash we talked about by returning early if no parse tree root exists.

Could you review this update and merge it if you approve? FYI @dschwen this also improves the VS Code integration.

Some examples of updated resource files causing the base document to be reprocessed is shown in the video below:

https://github.com/user-attachments/assets/1d856aa1-292b-4e48-b7f6-edfddaf2ea46

loganharbour commented:
For the sake of all of the updates, we can incorporate the wasp bump in https://github.com/idaholab/moose/pull/31479.

brandonlangley committed:
Fix MooseServer to properly format shorthand path notation syntax

refs #22766

loganharbour commented:
@brandonlangley drop your wasp updates in this PR. The submodule update is in master. Should be good to go after that change.

brandonlangley commented:
@loganharbour -

> drop your wasp updates in this PR. The submodule update is in master. Should be good to go after that change.

I just rebased on devel and dropped my WASP, conda, and versioner bumps. So this should be good to merge now.
